Refactored how begin/end/Job/Stream are implemented

Dr15Jones opened the pull request:
#### PR description:

- Moved transition calls from Worker to ModuleHolder
- The implementation is now done using stand-alone functions using ModuleRegistry
- Cleaned up some unused functions

The only changes expected are 
- some framework internal modules now have ActivityRegistry signals happen for them
- the order in which modules are called during begin/end/Job/Stream may have changed

This is part of the work to bring the event processing and EventSetup system closer in implementation. This is part of moving all non-data product related activities away from the Worker.
#### PR validation:

Code compiles. All framework unit tests pass.

resolves https://github.com/cms-sw/framework-team/issues/1431

Dr15Jones committed:
Removed commented out code

Dr15Jones committed:
Removed unused function workerType

Dr15Jones committed:
Add missing function implementation

-doBeginStream/doEndStream were declared but never defined for limited::OutputModuleBase

Dr15Jones committed:
Make end Job context message consistent

Dr15Jones committed:
Moved begin/endJob to ModuleHolder

As this is a one time call it makes more sense to associate with a module instead of a worker.

Dr15Jones committed:
Added ModuleRegistry::maxModuleID

Now keeps track of the maximum module ID.

Dr15Jones committed:
Move begin/endStream to ModuleHolder

Dr15Jones committed:
Do not need to pass ProcessContext to begin/endJob

The context is already known by most of the classes.

Dr15Jones committed:
GlobalContext needed for begin/endJob

The context is used to setup a thread-local via the ModuleContextSentry.

Dr15Jones committed:
Track only which modules fail beginJob/Stream

Dr15Jones committed:
Moved finishModulesInitialization to standalone function

Dr15Jones committed:
update to new begin/end/Job/Stream handling

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48416/45315

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/ExceptionThrowingProducer.cc  modified in PR(s): #44372

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48416/45315/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48416/45315/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Dr15Jones committed:
Fix formatting

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48416/45316

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/ExceptionThrowingProducer.cc  modified in PR(s): #44372

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones for master.

It involves the following packages:

- FWCore/Framework (**core**)
- FWCore/Integration (**core**)
- FWCore/ServiceRegistry (**core**)
- FWCore/TestProcessor (**core**)
- Mixing/Base (**simulation**)


@Dr15Jones, @civanch, @cmsbuild, @kpedro88, @makortel, @mdhildreth, @smuzaffar can you please review it and eventually sign? Thanks.
@fabiocos, @fwyzard, @makortel, @missirol, @wddgit this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented on FWCore/Framework/test/limited_producer_t.cppunit.cc:
```diff
@@ -402,20 +402,20 @@ testLimitedProducer::testLimitedProducer()
   //For each transition, bind a lambda which will call the proper method of the Worker
-  m_transToFunc[Trans::kBeginStream] = [](edm::Worker* iBase) {
+  m_transToFunc[Trans::kBeginStream] = [](edm::Worker* iBase, edm::maker::ModuleHolder* iHolder) {
     edm::StreamContext streamContext(s_streamID0, nullptr);
```
This variable should be removed

Dr15Jones replied:
done

wddgit commented:
I'll start reviewing this tomorrow.

Dr15Jones commented:
please test

Dr15Jones commented:
test parameters:
  - full = true

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 476KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1d8536/46925/summary.html
**COMMIT**: a5e25c665ba166e6b402332e98a0d89398509e00
**CMSSW**: CMSSW_15_1_X_2025-06-25-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48416/46925/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
Changes based on review

- removed unused variables
- updated a comment

Dr15Jones commented:
please abort test

cmsbuild commented:
Pull request #48416 was updated. @Dr15Jones, @civanch, @cmsbuild, @kpedro88, @makortel, @mdhildreth, @smuzaffar can you please check and sign again.

Dr15Jones commented:
test parameters:
  - full = true

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1d8536/46934/summary.html
**COMMIT**: 2c3bcc42237efc1d7e89e4f2daca2b0606b4f35c
**CMSSW**: CMSSW_15_1_X_2025-06-26-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48416/46934/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestPerfToolsModuleAllocMonitor had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 160 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-26-1100+1d8536/69144/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 10 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-26-1100+1d8536/69144/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4065619
 - DQMHistoTests: [Total failures: 51](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-26-1100+1d8536/69144/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4065548
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-26-1100+1d8536/69144/triggerResults): no differences found

Dr15Jones committed:
Update expected test output

Ordering of module calls in begin/end/Job/Stream changed.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48416/45332

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/ExceptionThrowingProducer.cc  modified in PR(s): #44372

wddgit reviewed (commented):
I got through it to the end of it and here are all the comments on things I noticed. Looks good in general. One additional comment. Can you add some explanation in the very top description of the PR as to why we are doing this? Just a couple sentences. Also reference an issue if there is one that is relevant.

wddgit commented on FWCore/Framework/src/ModuleRegistryUtilities.cc:
```diff
@@ -0,0 +1,179 @@
+          });
+          iActivityRegistry.preModuleBeginJobSignal_(holder->moduleDescription());
+          holder->beginJob();
+          failedSentry.release();
```
One additional thing, the comments in this function also apply to the next three analogous functions.

cmsbuild commented:
+1

**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1d8536/46939/summary.html
**COMMIT**: b00dec551c0656e9a8f6091025fef75c27a49113
**CMSSW**: CMSSW_15_1_X_2025-06-26-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48416/46939/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 159 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-26-1100+1d8536/69153/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 10 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-26-1100+1d8536/69153/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4065619
 - DQMHistoTests: [Total failures: 51](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-26-1100+1d8536/69153/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4065548
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-26-1100+1d8536/69153/triggerResults): no differences found

Dr15Jones committed:
Improvements based on review

- removed unnecessary includes
- removed redundant checks
- fixed comments

Dr15Jones committed:
Avoid possible exception while processing exception case

- Reuse ModuleSignalSentry for this case.
- Keep past behavior where exception in post signal during begin also resulted in module end not being called.

Dr15Jones commented on FWCore/Framework/src/Schedule.cc:
```diff
@@ -1153,21 +1154,19 @@ namespace edm {
                           eventsetup::ESRecordsToProductResolverIndices const& iESIndices,
                           ProcessBlockHelperBase const& processBlockHelperBase,
-                          ProcessContext const& processContext) {
-    auto const& processName = processContext.processConfiguration()->processName();
```
yes

Dr15Jones commented on FWCore/Framework/src/ModuleRegistryUtilities.cc:
```diff
@@ -0,0 +1,179 @@
+#include "FWCore/Framework/interface/ModuleRegistryUtilities.h"
+#include "FWCore/Framework/interface/ModuleRegistry.h"
+#include "FWCore/Common/interface/ProcessBlockHelperBase.h"
```
correct

Dr15Jones commented on FWCore/Framework/src/ModuleRegistryUtilities.cc:
```diff
@@ -0,0 +1,179 @@
+    });
+  }
+
+}  // namespace edm
```
added

Dr15Jones commented on FWCore/Framework/src/GlobalSchedule.cc:
```diff
@@ -80,29 +87,34 @@ namespace edm {
       } catch (cms::Exception& ex) {
-        exceptionContext(ex, globalContext, "Handling pre signal, likely in a service function");
-        throw;
+        if (!exceptionPtr) {
```
removed

Dr15Jones commented on FWCore/Framework/interface/ModuleRegistryUtilities.h:
```diff
@@ -0,0 +1,59 @@
+                              std::vector<unsigned int> const& beginJobFailedForModule) noexcept;
+
+}  // namespace edm
+#endif  // FWCore_Framework_ModuleRegistryUtilities_h
```
added

Dr15Jones commented on FWCore/Framework/interface/ModuleRegistryUtilities.h:
```diff
@@ -0,0 +1,59 @@
+
+  /** beginStreamFailedForModule holds module id for each module which threw an exception during
+     * the call to beginStream function. This vector is used to determine which modules should not
+     * have their endStream called. The vector should be passed to `runEndJobForModules`.
```
fixed

Dr15Jones commented on FWCore/Framework/src/ModuleRegistryUtilities.cc:
```diff
@@ -0,0 +1,179 @@
+          auto failedSentry = make_sentry(&beginJobFailedForModule,
+                                          [&](auto* failed) { failed->push_back(holder->moduleDescription().id()); });
+          auto sentry = make_sentry(&holder->moduleDescription(), [&](auto const* description) {
+            iActivityRegistry.postModuleBeginJobSignal_(*description);
```
I reworked the exception handling for all these functions to properly handle the case you identified.

cmsbuild commented:
Pull request #48416 was updated. @Dr15Jones, @civanch, @cmsbuild, @kpedro88, @makortel, @mdhildreth, @smuzaffar can you please check and sign again.

cmsbuild commented:
Pull request #48416 was updated. @Dr15Jones, @civanch, @cmsbuild, @kpedro88, @makortel, @mdhildreth, @smuzaffar can you please check and sign again.

wddgit commented:
+1 

Logic looks good. Just the one minor comment I just submitted. Overall, looks good now.

cmsbuild commented:
+1

**Size**: This PR adds an extra 56KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-1d8536/46983/summary.html
**COMMIT**: 2225e32c8677ee2b3db018da7b80511926a580d4
**CMSSW**: CMSSW_15_1_X_2025-06-27-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48416/46983/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 177 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-27-1100+1d8536/69183/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 8 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-27-1100+1d8536/69183/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4065619
 - DQMHistoTests: [Total failures: 73](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-27-1100+1d8536/69183/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4065526
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-27-1100+1d8536/69183/triggerResults): no differences found

civanch commented:
+1

Dr15Jones commented on FWCore/Framework/interface/maker/ModuleSignalSentry.h:
```diff
@@ -0,0 +1,62 @@
+#define FWCore_Framework_ModuleSignalSentry_h
+
+#include "FWCore/Utilities/interface/ConvertException.h"
+
```
I’ll add this change to another PR I’m working on in this package
